synapse/push/pusherpool.py

```diff
@@ -128,6 +127,22 @@ async def add_or_update_pusher(
         # stream ordering, so it will process pushes from this point onwards.
         last_stream_ordering = self.store.get_room_max_stream_ordering()
 
+        # Before we actually persist the pusher, we check if the user already has one
+        # for this app ID and pushkey. If so, we want to keep the access token and
+        # device ID in place, since this could be one device modifying
+        # (e.g. enabling/disabling) another device's pusher.
+        # XXX(quenting): Even though we're not persisting the access_token_id for new
+        # pushers anymore, we still need to copy existing access_token_ids over when
+        # updating a pusher, in case the "set_device_id_for_pushers" background update
+        # hasn't run yet.
+        access_token_id = None
+        existing_config = await self._get_pusher_config_for_user_by_app_id_and_pushkey(
+            user_id, app_id, pushkey
+        )
+        if existing_config:
+            device_id = existing_config.device_id
+            access_token_id = existing_config.access_token
```

Comment:
I think this causes a race with the background update:

1. We fetch an existing pusher config where `device_id` is `None`, but `access_token_id` is set.
2. The background update processes this row, adding the `device_id`.
3. We call `add_pusher` and overwrite the `device_id` back to `None`.

Is that possible?

Reply:
I guess, but it is very unlikely, because unless it happens at the same time as the very last batch of the background update, it will be cleared right after by it
Also, this would only impact old pushers where we don't have the device_id inserted
So, it would have to be at the same time an old enough pusher that doesn't already have a device_id, and a recent enough pusher to be included in the very last batch of background updates

Reply:
We talked about this and it *should* be OK since when we add the pusher back (step 3) it actually gets a new ID. So this is an extremely unlikely event to occur.